pypeit/coadd2d.py

```diff
@@ -169,39 +173,117 @@ def __init__(self, spec2d, spectrograph, par, det=1, offsets=None, weights='auto
+
+        # Check that binning is the same for all the exposures
+        binspec_list = [slits.binspec for slits in self.stack_dict['slits_list']]
+        binspat_list = [slits.binspat for slits in self.stack_dict['slits_list']]
+        if not len(set(binspec_list)) == 1:
+            msgs.error('Not all of your exposures have the same spectral binning. Check your inputs')
+        if not len(set(binspat_list)) == 1:
+            msgs.error('Not all of your exposures have the same spatial binning. Check your inputs')
         self.binning = np.array([self.stack_dict['slits_list'][0].binspec,
                                  self.stack_dict['slits_list'][0].binspat])
+
         self.spat_ids = self.stack_dict['slits_list'][0].spat_id
 
         # If smoothing is not input, smooth by 10% of the spectral dimension
         self.sn_smooth_npix = sn_smooth_npix if sn_smooth_npix is not None else 0.1*self.nspec
 
-        # maskdef info
-        self.maskdef_id = np.array([slits.maskdef_id for slits in self.stack_dict['slits_list']])
-        self.maskdef_offset = np.array([slits.maskdef_offset for slits in self.stack_dict['slits_list']])
-        self.maskdef_objpos = np.array([slits.maskdef_objpos for slits in self.stack_dict['slits_list']])
-        self.maskdef_slitcen = np.array([slits.maskdef_slitcen for slits in self.stack_dict['slits_list']])
+    @property
+    def nslits_coadded(self):
+        """
+        This gives the number of slits of the coadded frame.
+        `self.good_slits` must be defined first.
+
+        Returns:
+            :obj:`float`: Number of slits of the coadded frame
+
+        """
+        # this is the number of slits of the coadded frame
+        if self.good_slits is None:
+            msgs.error('Cannot determine nslits_coadded because `self.good_slits` is `None`. '
+                       'Run `self.good_slits = self.good_slitindx(only_slits=only_slits)` first')
+        else:
+            return self.good_slits.size
+
+    def good_slitindx(self, only_slits):
+        """
+        Args:
+            only_slits (:obj:`list`, optional):
+                List of slits to combine. It must be `slitord_id`
+
+        Returns:
+            `numpy.ndarray`_: array of index of good slits in the un-coadded frames
+        """
+        # TODO We should be checking the bitmask for the reductions or something here??
+        #  Yes, definitely - DONE?
+
+        only_slits = [only_slits] if (only_slits is not None and
+                                      isinstance(only_slits, (int, np.int, np.int64, np.int32))) else only_slits
+
+        # This creates a unified bpm common to all frames
+        slits0 = self.stack_dict['slits_list'][0]
+        # bpm for the first frame
+        reduce_bpm = (slits0.mask > 0) & (np.invert(slits0.bitmask.flagged(slits0.mask,
+                                                                           flag=slits0.bitmask.exclude_for_reducing)))
```

Comment:
Hi @kbwestfall ! Thanks for this comments. I remember spending a lot of time in the past trying to figure out how bitmask works and I don't think I still fully understand it. But what you says does not work here because of the following example:
```
In [17]: slits.mask
Out[17]: array([ 0, 82,  0], dtype=int16)

In [18]: (slits.mask > 0) & (np.invert(slits.bitmask.flagged(slits.mask, flag=slits.bitmask.exclude_for_reducing)))
Out[18]: array([False,  True, False])

In [19]: slits.mask > 0
Out[19]: array([False,  True, False])

In [20]: np.invert(slits.bitmask.flagged(slits.mask, flag=slits.bitmask.exclude_for_reducing))
Out[20]: array([ True,  True,  True])
```
In this case I definitely want to mask the middle slit and the combination of the conditions above gives me the right mask, as it does only `slits.mask > 0`, but the second condition with `slits.bitmask.flagged` alone does not give me the right masking. Maybe it is the second one that is redundant, but there may be slits with one of the flags specified in `slits.bitmask.exclude_for_reducing` that we don't want to mask. I don't know but it's very confusing.

Reply:
Thanks for the detailed response!  And sorry for not realizing you were inverting the second selection.  In that case, the second statement is finding all bits not masked with the `exclude_for_reducing` flags, which includes bits that are not masked at all.

If I understand correctly, you're trying to find all the slits flagged by anything *except* the flags in `exclude_for_reducing`.  Of course, the way you've done it works, but you could also create the relevant list of flags, which is admittedly requires some digging to figure out how to do.  I'm fine with you leaving this as is, but just for reference, here's how to do it with a single call to `flagged`:

```
>>> from pypeit.slittrace import SlitTraceBitMask
>>> bm = SlitTraceBitMask()
>>> import numpy as np
>>> mask = np.array([0, 82, 0], dtype=np.int16)
>>> np.setdiff1d(bm.keys(), bm.exclude_for_reducing)
array(['BADFLATCALIB', 'BADREDUCE', 'BADTILTCALIB', 'BADWVCALIB',
       'BOXSLIT', 'SHORTSLIT', 'USERIGNORE'], dtype='<U13')
>>> bm.flagged(mask, flag=np.setdiff1d(bm.keys(), bm.exclude_for_reducing))
array([False,  True, False])
```